Support capturing rotated displays on Windows

ns6089 opened the pull request:
## Description
<!--- Please include a summary of the changes. --->

Based on https://github.com/LizardByte/Sunshine/pull/1560 to prevent future merge conflicts.

Portrait mode displays currently can't be captured at all, this PR fixes it. Also fixes landscape (flipped) being upside down.

Streaming resolution is still decided by Moonlight.

### Screenshot
<!--- Include screenshots if the changes are UI-related. --->

### Issues Fixed or Closed
<!--- Close issue example: `- Closes #1` --->
<!--- Fix bug issue example: `- Fixes #2` --->
<!--- Resolve issue example: `- Resolves #3` --->
- Fixes #1292 

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

ns6089 committed:
Support display rotation on Windows

ns6089 committed:
Use different slots for different rotations

So we won't get hard to debug errors later when we decide to support
serial processing on Windows for example.

ns6089 commented on src/platform/windows/display_vram.cpp:
```diff
@@ -993,8 +1008,29 @@ namespace platf::dxgi {
+        switch (display_rotation) {
+          case DXGI_MODE_ROTATION_UNSPECIFIED:
+          case DXGI_MODE_ROTATION_IDENTITY:
+            cursor.set_pos(pointer_position.Position.x, pointer_position.Position.y, pointer_position.Visible);
```
This part may need deeper changes in `set_pos()` itself, if cursor image is not square (can this even happen?).

ns6089 replied:
Pushed a fix, even though all the cursors I've received from the api so far were square sized.

cgutman commented on src_assets/windows/assets/shaders/directx/SceneVS.hlsl:
```diff
@@ -1,22 +1,37 @@
+        float2x2 rotation_matrix = { cos(rotation_radians), -sin(rotation_radians),
+                                     sin(rotation_radians), cos(rotation_radians) };
+        float2 rotation_center = { 0.5, 0.5 };
+        output.tex = round(rotation_center + mul(rotation_matrix, output.tex - rotation_center));
```
Can we precompute these values and put them in the constant buffer?

ns6089 replied:
We can, but this is in the vertex shader that runs for 3 vertices. So maybe better not prematurely optimize it. Same reason why I replaced bit-shift magic with easier to understand `if` branches.

cgutman replied:
Normally I'd agree, but in this case we're already creating the constant buffer to store the rotation angle and the matrix logic ends up getting duplicated in several shaders. It seems like moving it to C++ would be cleaner and faster.

ns6089 replied:
That's two separate problems, we can rule "faster", 5 basic math functions called for 3 vertices two or three times per frame is literally nothing for gpu.

Now for the duplication, this one is legitimate concern. But if we move the rotation to generic matrix defined externally to the shader, it will start implying that we support arbitrary angle rotation and rounding part will stop making sense.
`output.tex = round(rotation_center + mul(rotation_matrix, output.tex - rotation_center));`
So maybe we can do de-duplication in shader? By using some `common.hlsl` and including it around.

cgutman replied:
Ok, I don't feel very strongly about this. If you'd much rather leave it as is or change it to use an HLSL include to deduplicate the code, either is fine with me.

ns6089 replied:
Can this be done in separate pull request?
There are more things that can be deduplicated other than this rotation matrix, might as well tackle all of them.

cgutman replied:
Yeah that's fine with me.

cgutman commented on src/platform/windows/display_vram.cpp:
```diff
@@ -938,6 +953,10 @@ namespace platf::dxgi {
     }
 
+    if (t.Width != t.Height) {
+      BOOST_LOG(error) << "NOT SQUARE";
```
If this isn't just a debugging artifact that will be removed, change the message to something better.

ns6089 replied:
Oops. my bad.

ns6089 committed:
nvenc: generate less IDR frames if rfi is delayed

If IDR frame was generated due too rfi request being too large, we
didn't save it in last rfi range.

Also move rfi_needs_confirmation variable modification to a more
logical place, even if doesn't affect how current code works.
Any unfulfilled rfi request needs confirmation, whether it's direct rfi
flag in transported frame or indirect IDR frame.

Vanzelote commented:
I have Windows 11 22H2 ARM64 installed on a Pocophone F1 (Snapdragon 845). I also installed the Sunshine-Windows-ARM64-installer.msi version, but I'm not getting horizontal screen capture when rotating the Windows screen. I've already tried locking screen rotation in landscape mode and starting the connection, but everything indicates that Sunshine insists on the device's original resolution of 1080 x 2246. Perhaps Moonlight could also be the problem? The Moonlight that appears in the background of the video is running on a Samsung DeX connected to an LG 24" FHD monitor. I've already tested it on other Moonlight clients and I get the same problem. Can you help?

https://github.com/user-attachments/assets/f64ea54d-5794-4d31-9262-ac261a9ea5db
